### ISSUE TITLE
Memory/data corruption / crash on Lenovo T440p (GT 730M).

### ISSUE THREAD

**Issue by leoluk** [Bumblebee-Project/bbswitch#78]
I just installed bbswitch on the newly-released Thinkpad T440p.

Loading the bbswitch module and disabling the card works perfectly fine:

```
[  142.881587] bbswitch: version 0.7 
[  142.881593] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.VID_
[  142.881596] bbswitch: Found discrete VGA device 0000:02:00.0: \_SB_.PCI0.PEG_.VID_
[Package] (20130517/nsarguments-95)  
[  142.882097] bbswitch: detected an Optimus _DSM function
[  142.882106] pci 0000:02:00.0: enabling device (0004 -> 0007)
[  142.882127] bbswitch: Succesfully loaded. Discrete card 0000:02:00.0 is on
[  156.250136] bbswitch: disabling discrete graphics
[Package] (20130517/nsarguments-95)  
[  156.265409] thinkpad_acpi: EC reports that Thermal Table has changed
[  156.376985] pci 0000:02:00.0: power state changed by ACPI to D3cold
```

But enabling it again seems to mess up the PCI bus, crash the network adapters, and cause data corruption (files read from the disk contain random characters, filesystem errors, some files are missing, empty or filled with random data after a reboot).

```
[  160.406244] bbswitch: enabling discrete graphics
[  160.647323] pci 0000:02:00.0: power state changed by ACPI to D0 
[  160.647336] thinkpad_acpi: EC reports that Thermal Table has changed
```

There are no bbswitch errors, but shortly after entering the command, the syslog fills with various kernel messages related to internal devices no longer responding. The filesystem sometimes remounts as read-only, and the system becomes unusable and has to be reset.

At this point, the machine is unable to write files to the disk or a USB stick or communicate with the network, so I made some "screen shots" using my smartphone. I tried to redirect the syslog to another machine using the internal network as well as a USB WLAN adapter, but the data cuts off as soon as the graphics card is enabled.

Installing the other bumblebee components or the Nvidia/Nouveau drivers does not make any difference.

**Comment by ksonj** [comment]
Similar problem on a T440p, BIOS v2.3, Linux Kernel 3.19 (with ck patches). I can use the card with bumblebee without issues, I cannot however switch it off via bbswitch. With adding acpi_osi='!Windows 2013' to the kernel options, the card is disabled after booting, re-eanbling it leads to already mentioned filesystem corruptions, though.

**Comment by flurri** [comment]
I tried out @doudou 's patch and used nouveau, and found that the same issue happens (instead of using Bumblebee, I've been using PRIME):

```
[   51.010655] bbswitch: discrete card is in state D3
[   51.010656] bbswitch: discrete card is in state D3
[   51.010656] bbswitch: enabling discrete graphics
[...]
[   54.215695] bbswitch: discrete card is in state D0
[   54.215697] bbswitch: device 0000:02:00.0 is in use by driver 'nouveau', refusing OFF
```
But even though it refuses OFF, `cat /proc/acpi/bbswitch` returns `OFF`. Also, watching powertop, I see very little to no change in power output (and what I am seeing I'm assuming is probably more due to processor scaling than anything). On the bright side, I haven't noticed any power issues and when starting up my system with the card off, it seems to actually be off:
```
[    2.188988] bbswitch: version 0.8
[    2.188994] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.VID_
[    2.188999] bbswitch: Found discrete VGA device 0000:02:00.0: \_SB_.PCI0.PEG_.VID_
[    2.189274] bbswitch: detected an Optimus _DSM function
[    2.189278] bbswitch: discrete card is in state D0
[    2.189312] bbswitch: discrete card is in state D0
[    2.189313] bbswitch: disabling discrete graphics
[    2.201842] bbswitch: discrete card is in state D3
[    2.201844] bbswitch: Succesfully loaded. Discrete card 0000:02:00.0 is off
```
 Not sure if there's any definite way to tell, though.

**Comment by doudou** [comment]
> Not sure if there's any definite way to tell, though.

I'm using lspci -v to check that the card is in D3 state. I unfortunately did not manage to get a reliable D-state report from the kernel, so sometimes bbswitch will report D3 while lspci D0.

**Comment by flurri** [comment]
After trying to figure out why nouveau was started right away instead of stopping, I found that I forgot to blacklist the module. The change between the two seems to be a setting of "latency 0", as I get this at startup:
```
02:00.0 VGA compatible controller: NVIDIA Corporation GK208M [GeForce GT 730M] (rev a1) (prog-if 00 [VGA controller])
	Subsystem: Lenovo Device 221d
	Flags: fast devsel, IRQ 17
	Memory at f0000000 (32-bit, non-prefetchable) [size=16M]
	Memory at c0000000 (64-bit, prefetchable) [size=256M]
	Memory at d0000000 (64-bit, prefetchable) [size=32M]
	I/O ports at 3000 [size=128]
	Expansion ROM at <ignored> [disabled]
	Capabilities: [60] Power Management version 3
	Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [78] Express Endpoint, MSI 00
	Capabilities: [100] Virtual Channel
	Capabilities: [128] Power Budgeting <?>
	Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
	Capabilities: [900] #19
	Kernel modules: nouveau
```
and this after sending `ON` to `/proc/acpi/bbswitch`:
```
02:00.0 VGA compatible controller: NVIDIA Corporation GK208M [GeForce GT 730M] (rev a1) (prog-if 00 [VGA controller])
	Subsystem: Lenovo Device 221d
	Flags: bus master, fast devsel, latency 0, IRQ 17
	Memory at f0000000 (32-bit, non-prefetchable) [size=16M]
	Memory at c0000000 (64-bit, prefetchable) [size=256M]
	Memory at d0000000 (64-bit, prefetchable) [size=32M]
	I/O ports at 3000 [size=128]
	Expansion ROM at <ignored> [disabled]
	Capabilities: [60] Power Management version 3
	Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [78] Express Endpoint, MSI 00
	Capabilities: [100] Virtual Channel
	Capabilities: [128] Power Budgeting <?>
	Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
	Capabilities: [900] #19
	Kernel modules: nouveau
```
Flipping between `ON` and `OFF` changes that, and no errors are reported in dmesg. That said, blacklisting nouveau has the effect of also losing the ability to actually use the nvidia card (xrandr only reports the Intel chip), so I'm not sure what to do by that point, as even manually running modprobe for nouveau changes nothing.

**Comment by ArchangeGabriel** [comment]
@flurri Indeed, for PRIME you actually need the main X.org server to be started with nouveau loaded.

**Comment by flurri** [comment]
@ArchangeGabriel Ah, my mistake then. I forget what my reason was for not using bumblebee, but in any case, on Kubuntu 15.04 with bumblebee from APT, the nouveau module is loaded/unloaded as needed, although I still get the `Refused to change power state, currently in D0` error. Also, the 'bus master' and 'latency 0' options seem to be enabled/disabled as the card is (presumably) turned on/off, respectively.

**Comment by ArchangeGabriel** [comment]
The `Refused to change power state, currently in D0` error is harmless.

**Comment by nyim** [comment]
I having the same memory corruption,
lspci , lshw crash even in virtual tty(Ctrl+Alt+F2).

bios version 1.18.

arch linux kernel 4.1.4 or 2015.7.1 install iso.

Is this the only solution flashing bios to 1.14?

disable modeset cause arch linux fail to load Display Manager, but lspci/lshw work fine in virtual tty

**Comment by leoluk** [comment]
Did anyone try to disable VT-d (not VT-x!) and see if the problem persists?

**Comment by flurri** [comment]
The Thinkpad T440p I have (Core i7-4700MQ) doesn't have VT-d. I'm not sure if there are any other models that do, though.

**Comment by ArchangeGabriel** [comment]
Pull request https://github.com/Bumblebee-Project/bbswitch/pull/102 is working around bbswitch reading PCI configuration space potentially being a mistake.

This could also probably be related to #112 and `ACPI_OSI` issues, in particular the need to handle `_DSM` differently in some cases and the potential need to hande PCI-E link PM.

**Comment by Lekensteyn** [comment]
Can someone affected by the memory corruption bug test these patch series based on Linux v4.6:
https://lekensteyn.nl/files/linux-v4.6-pcipm-nouveau-pm2.patch
(git branch at https://git.lekensteyn.nl/peter/linux/log/?h=nouveau-pm2)

Four of the patches are for the PCI core and will likely end up in v4.8. Four others are for nouveau and is currently under review. With these patches, the DSM method is not called anymore and parent PCIe port is enabled/disabled via runtime PM.

Please disable bbswitch and nvidia blob before testing and then load nouveau. After five seconds, the GPU will sleep via runtime PM. Invoking "lspci" will enable it again. Does suspend/resume still work? Any observable (absence of) corruption?

**Comment by vpoughon** [comment]
@Lekensteyn could you link to threads where you submitted those patches? I don't think I can't try it with my machine but at least I could monitor their progress and upgrade to 4.8 release when it comes out.

**Comment by dionorgua** [comment]
@Lekensteyn, just tried it on affected t440p with latest BIOS (2.37. Lenovo says something about Win10 support in changelog).
  
Using kernel 4.6.1, patched with PSR fix for intel and your linux-v4.6-pcipm-nouveau-pm2.patch for nvidia. No bbswitch, no NVIDIA blob. 

Got followed ACPI warns in dmesg:

[   53.568692] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160108/nsarguments-95)
[   53.570543] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160108/nsarguments-95)
[   53.571820] iwlwifi 0000:04:00.0: loaded firmware version 16.242414.0 op_mode iwlmvm
[   53.577239] pstore: Registered efi as persistent store backend
[   53.578242] pci 0000:02:00.0: optimus capabilities: enabled, status dynamic power, hda bios codec supported
[   53.578247] VGA switcheroo: detected Optimus DSM method \_SB_.PCI0.PEG_.VID_ handle
[   53.578247] nouveau: detected PR support, will not use DSM
[   53.578508] nouveau 0000:02:00.0: enabling device (0004 -> 0007)
[   53.578560] nouveau 0000:02:00.0: NVIDIA GK208 (108100a1)

I've just called powertop --auto-tune to make sure that runtime PM is used. And it works!. lspci wakes up nvidia card and I'm getting followed messages:

wakeup:
[  635.620999] thinkpad_acpi: EC reports that Thermal Table has changed
[  635.724142] nouveau 0000:02:00.0: DRM: resuming kernel object tree...
[  635.766949] nouveau 0000:02:00.0: priv: HUB0: 085014 ffffffff (1f70820b)
[  635.766956] nouveau 0000:02:00.0: priv: GPC0: 4188ac 00000001 (1a70822e)
[  635.809202] nouveau 0000:02:00.0: DRM: resuming client object trees...

suspend:
[  640.805535] nouveau 0000:02:00.0: DRM: evicting buffers...
[  640.805539] nouveau 0000:02:00.0: DRM: waiting for kernel channels to go idle...
[  640.805556] nouveau 0000:02:00.0: DRM: suspending client object trees...
[  640.806797] nouveau 0000:02:00.0: DRM: suspending kernel object tree...
[  641.274023] thinkpad_acpi: EC reports that Thermal Table has changed

but lspci now takes ~0.6secs when it needs to wakeup card. and 0.2sec when card is still active.

s2ram works. I've not checked that NVIDIA card actually works. Can try later if you need it.

Btw I've found followed new ACPI event when nvidia card is suspended/resumed:
ibm/hotkey LEN0068:00 00000080 00006030

I'll keep this kernel running on my laptop for a while to make sure that everything is ok.

**Comment by dionorgua** [comment]
Full dmesg after a few s2ram cycles:
[dmesg_461_pm.txt](https://github.com/Bumblebee-Project/bbswitch/files/297426/dmesg_461_pm.txt)

**Comment by dionorgua** [comment]
Btw. I've tried to use nouveau using DRI_PRIME=1 and it works! Card is waked up and suspended when needed.

Any chance that NVIDIA blob will work?

**Comment by Lekensteyn** [comment]
@vpoughon For some reason only a cover letter + two patches (instead of five) show up for v5 of the PCI patches, see http://article.gmane.org/gmane.linux.power-management.general/76872. The tree with v4 is available at https://git.kernel.org/cgit/linux/kernel/git/helgaas/pci.git/?h=pci/pm

For v1 of the nouveau patches, see https://lists.freedesktop.org/archives/nouveau/2016-May/025116.html (a v2 has not been sent yet, the current branch in git will form v2).

**Comment by Lekensteyn** [comment]
@dionorgua Thank you for testing! The warnings about the DSM parameter type are harmless, it is a pecularity of the firmware that deviates from the ACPI standard. Using `powertop --autotune` is possible, but should not be needed for RPM with nouveau and the pcie port (this is done automatically by the drivers).

`lspci` takes longer because it has to wake up the device first before it can read the PCI space. This also means that you can no longer rely on `lspci -d10de: | grep "rev ff"` to check for the power state.

The blob does not use runtime PM so these patches are unlikely to help in that case. If possible I suggest you to stick to nouveau for power saving, a bit of acceleration and external monitor support. A more complete update for bbswitch should be ready for testing in a month, until then you can use these patches and unload nouveau if you do not want to save power anymore and start using the blob. Note: if you plan to unload nouveau, be sure to apply these two fixes (the second is needed, but the first one changes the context):
https://lists.freedesktop.org/archives/nouveau/2016-May/025107.html
https://lists.freedesktop.org/archives/nouveau/2016-May/025108.html

**Comment by dionorgua** [comment]
Thanks a lot!. I was using acpi_call before to turn off NVIDIA card forever. Now I'll stick with just loading nouveau. It still works well!.

Also as far as I see NVIDIA card is not needed for external monitors (at least on t440p). Everything (including HDMI/DisplayPort on dock) works by using intel. It looks like NVIDIA card has no displays attached.

**Comment by dionorgua** [comment]
Ok. Now I'm running this patched kernel for a bit more time. It's pretty stable. Everything works as expected.

Also In the past I've had 'nouveau.modeset=0' in my cmdline. And good thing is that even with this option power management works. At least it disables NVIDIA card.

**Comment by syddel** [comment]
Hi. I've read the posts in this thread but a lot of it is going straight over my head. I have a P65_P67RGRERA laptop which has the Intel i7-6820HK CPU, Intel 530 GPU, and nVidia GTX 970m. I have installed Ubuntu 16.04, the latest 4.6.2 kernel, and the nVidia 364 driver from ppa:graphics-drivers/ppa.

When I go in to my BIOS and switch the graphics to "MSHYBRID" (from "DISCRETE"), everything seems to work (except for the HDMI/DisplayPort), and the power consumption is high (around 27 Watts when idle).

I managed to turn off the power to the nVidia GPU using:

```
   rmmod nvidia_drm nvidia_modeset nvidia
    tee /proc/acpi/bbswitch <<<OFF
```

After running the above, power consumption drops to around 13 Watts when idle, but whenever I run lspci twice, or lshw, my laptop instantly freezes (holding down the power button is the only way to restart the machine).

I would really like to get everything working in Intel graphics mode. Is it fairly straight forward to install the patches mentioned in this thread? Is there anything else I can do to get Intel Graphics mode running without crashes (and with the mini DisplayPort/HDMI)? Should I just wait for the 4.8 kernel?

**Comment by dionorgua** [comment]
Hi, It depends... On my Lenovo t440p machine all outputs (VGA, miniDP plus all ports on Dock) are wired to Intel GPU. I mostly don't care about graphics performance, so it's enough to just turn off NVIDIA graphics forever.

It's be possible that your laptop is different and HDMI/DP connectors are wired to NVIDIA card. In such case you can't get Intel-only setup. I think you should somehow check it first. I suggest to completely uninstall NVIDIA blob and take a look to xrandr -q output.

As about bbswitch, as far as I understand it's not safe to use it on latest laptops at all. 

If you don't care about performance, probably you can try 4.6.x kernel with patch posted by @Lekensteyn for now (or wait for kernel 4.8) and nouveau driver. First of all get patched kernel working, make sure that runtime PM works and then read 'Reverse PRIME' section here:  https://nouveau.freedesktop.org/wiki/Optimus/

**Comment by Pitometsu** [comment]
Will it work with proprietary nvidia drivers? Also are there anybody who experienced remounting filesystem as read-only after suspend? And also got error `2100: Detection error on HDD0 (Main HDD)` after each reboot with not properly unmounted main hdd. Is there any known solution for this?

**Comment by Lekensteyn** [comment]
@syddel boot with `acpi_osi="!Windows 2015"` on the P66xRx/P67xRx family of laptops, these have a Windows 10-specific firmware thing that sometimes (actually, quite often) causes issues whenever you are resuming the nvidia graphics card.

@Pitometsu bbswitch works with proprietary drivers, nouveau could possibly also be used (if you unload nvidia). It does not sound normal to have fs corruption after suspend... if you have a different laptop and a bbswitch issue, please open a new issue.

**Comment by Pitometsu** [comment]
@Lekensteyn I have T440p

**Comment by syddel** [comment]
@Lekensteyn - Thank you! I can now run lspci and lshw in Intel graphics mode without the laptop completely freezing up. I no longer get random freezing during normal use, and my laptop is using about 7 Watts less in nVidia mode, and about 2 Watts less in Intel mode. I can also shutdown or reset the laptop without it freezing.

External monitors still don't work when using Intel graphics, but I've read that this may be because the display outputs are wired directly to the nVidia card.

**Comment by Pitometsu** [comment]
@syddel how get all this fixes? New kernel version? New bumblebee? Something patched?

**Comment by Lekensteyn** [comment]
@Pitometsu You need an updated kernel. Mainline (v4.7-rc) will not have the fixes yet, see comment https://github.com/Bumblebee-Project/bbswitch/issues/78#issuecomment-223072012 for the patches.

**Comment by dionorgua** [comment]
Just for the record. https://lekensteyn.nl/files/linux-v4.6-pcipm-nouveau-pm2.patch still applies correctly to 4.7 kernel. And works as expected.

**Comment by Lekensteyn** [comment]
See the above message? Good news, this will be fixed in Linux 4.8 if everything goes well. :smiley:
It is queued in this branch: https://github.com/skeggsb/linux/commits/linux-4.8

**Comment by dionorgua** [comment]
fingers crossed

**Comment by dionorgua** [comment]
It seems everything works out of the box with kernel 4.8!

**Comment by kubik369** [comment]
How is the battery life? :)

**Comment by kubik369** [comment]
@dionorgua Have you encountered any issues since you updated? :)

**Comment by dionorgua** [comment]
Hi,

Running 4.8 for a while right now (without any patches). Everything seems to be OK. But I don't use NVIDIA GPU at all. 
Tried to launch a few games using DRI_PRIME=1. And it works (card is resumed when needed and then goes to sleep once it isn't needed).

**Comment by kubik369** [comment]
Does the left side of the keyboard still get hot or is it okay when the
Nvidia GPU is turned off?

On Fri, 14 Oct 2016 16:33 Dmitry Nezhevenko, <notifications@github.com>
wrote:

Hi,

Running 4.8 for a while right now (without any patches). Everything seems
to be OK. But I don't use NVIDIA GPU at all.
Tried to launch a few games using DRI_PRIME=1. And it works (card is
resumed when needed and then goes to sleep once it isn't needed).

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Bumblebee-Project/bbswitch/issues/78#issuecomment-253817464>,
or mute the thread

**Comment by dionorgua** [comment]
It's became a bit hot when I have some CPU intensive tasks. But in normal conditions (like browsing) it's just slightly warm.

**Comment by xqms** [comment]
I still see memory corruption using kernel 4.8.6 on T440p. Turning the card off works, but if I run `lspci` or start Xorg (which apparently also enumerates the card), `lspci` freezes and I get random failures in dmesg and/or filesystem corruption.

The last message from nouveau is "DRM: resuming kernel object tree..."

@dionorgua Are you using any special kernel cmdline parameters like apci_osi=XYZ? Anything else that might be special in your setup?

I'm running BIOS version 1.17, which is pretty old. Should I update?

**Comment by Lekensteyn** [comment]
@xqms Are you using any special kernel parameters or modules? Can you post your full dmesg somewhere?

**Comment by xqms** [comment]
@Lekensteyn I have no custom kernel modules. I'm using the ubuntu mainline 4.8.6 kernel, which is pretty vanilla.

I will try to capture dmesg, although there is a bit of luck involved whether the system stays alive long enough to save something.

**Comment by xqms** [comment]
dmesg is here: https://gist.github.com/xqms/9c5f35509a4ea9e2d9a6be9bb55c50b7

This time the system froze completely without further evidence for memory corruption.

**Comment by dionorgua** [comment]
My parameters from /proc/cmdline:
`BOOT_IMAGE=/vmlinuz-4.8.5+ root=/dev/mapper/debpad2-root ro elevator=deadline init=/lib/sysvinit/init intel_iommu=on init=/bin/systemd i915.enable_psr=0`

I don't think that there is something really special here. i915.enable_psr=0 is needed to avoid display flickering when external monitor is connected.  

As about BIOS, I currently use 2.37. I think that it's better to update (because there were 'Win8/Win10 support entries in changelog that are probably related to Optimus'. Probably old BIOS is root cause of memory corruption for you. There are also some differences in dmesg. Your:
```
[   11.580024] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160422/nsarguments-95)
[   11.580226] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160422/nsarguments-95)
[   11.580489] pci 0000:02:00.0: optimus capabilities: enabled, status dynamic power, hda bios codec supported
[ 11.580492] VGA switcheroo: detected Optimus DSM method \_SB_.PCI0.PEG_.VID_ handle
```

My:
```
Oct 27 18:56:17 debpad kernel: [  140.650174] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160422/nsarguments-95)
Oct 27 18:56:17 debpad kernel: [  140.650498] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160422/nsarguments-95)
Oct 27 18:56:17 debpad kernel: [  140.650805] pci 0000:02:00.0: optimus capabilities: enabled, status dynamic power, hda bios codec supported
Oct 27 18:56:17 debpad kernel: [  140.650809] VGA switcheroo: detected Optimus DSM method \_SB_.PCI0.PEG_.VID_ handle
Oct 27 18:56:17 debpad kernel: [  140.650809] nouveau: detected PR support, will not use DSM
```

So you don't have message about PR and 'will not use DSM'.

But in any case I just want to warn you that once updated, there will be no way to downgrade back.

**Comment by Lekensteyn** [comment]
@dionorgua Could you upload your acpidump? Alternatively, upload acpidump+dmidecode+lspci following the instructions at https://bugs.launchpad.net/lpbugreporter/+bug/752542

In a slightly older tarball for this laptop (LENOVO-20AWS02A00, BIOS GLET41WW (1.16), 10/27/2013), there seems to be some code for Windows 8 compatibility, but I cannot judge whether it is good or bad. Having something to compare would be nice.

**Comment by xqms** [comment]
I uploaded my data here: https://bugs.launchpad.net/lpbugreporter/+bug/752542/comments/801

I'll try the BIOS update next.

**Comment by xqms** [comment]
Yep, the BIOS update did the trick. `lspci` no longer freezes the system and so far I had none of the memory corruption symptoms.

acpidump from the new BIOS: https://bugs.launchpad.net/lpbugreporter/+bug/752542/comments/802

**Comment by dionorgua** [comment]
It's from BIOS 2.37.
[LENOVO-20ANCTO1WW.tar.gz](https://github.com/Bumblebee-Project/bbswitch/files/570853/LENOVO-20ANCTO1WW.tar.gz)

**Comment by dionorgua** [comment]
@xqms you've updated to 2.39, everything else is OK?

**Comment by xqms** [comment]
@dionorgua Yes, so far everything works. I haven't done anything fancy with the nvidia GPU yet, though.

**Comment by Lekensteyn** [comment]
Hm, BIOS 1.17 has a date before 2015 (11/14/2013), so that is why the new PR3 functionality in kernel 4.8 was not activated. Between 1.17 and 2.39 there are no significant changes in the ACPI tables that could have an influence on this.

**Comment by abbradar** [comment]
@Lekensteyn,

> @dionorgua on the latest laptops, bbswitch is not necessary unsafe but it might not fully turn your graphics card off.

Do I understand correctly that currently there's no way to have both proper PM (fully turned off card) and nvidia proprietary drivers at all? I'm not complaining, it's just that there are several threads with discussions mostly about nouveau going on, and it's difficult to catch up on the progress.

**Comment by dionorgua** [comment]
Probably no. As far as I understand currently (with latest kernel and DRI_PRIME), nouveau is responsible for power management of NVIDIA card. 

It's maybe possible to get power management + NVIDIA driver using dedicated X server by using some scripts that will rmmod nvidia && modprobe nouveau once external GPU is unused. But I don't know whether it's implemented somewhere or not.

**Comment by abbradar** [comment]
Yeah, I've thought about something along those lines if bbswitch is not a full-featured option currently. Thanks for the confirmation!

**Comment by Lekensteyn** [comment]
If you are feeling adventurous, there is a pm-rework branch which you can try. The vgaswitcheroo addition is however not stable, it oopsed somewhere. That is, try the branch at https://github.com/Bumblebee-Project/bbswitch/commit/5c7b3f53f229c70bc49c710295967605ac5846e4

**Comment by abbradar** [comment]
I'd be glad to. Do you want any specific tests or just feedback on whether it works? 

On November 20, 2016 11:57:20 PM GMT+03:00, Peter Wu <notifications@github.com> wrote:
>If you are feeling adventurous, there is a pm-rework branch which you
>can try. The vgaswitcheroo addition is however not stable, it oopsed
>somewhere. That is, try the branch at
>https://github.com/Bumblebee-Project/bbswitch/commit/5c7b3f53f229c70bc49c710295967605ac5846e4
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Bumblebee-Project/bbswitch/issues/78#issuecomment-261805117

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Lekensteyn** [comment]
@abbradar Some people have reported that it works for them, maybe it is also good enough for you. YMMV

**Comment by abbradar** [comment]
Lenovo T440p, BIOS 2.39. Disconnected external displays, cold boot, bumblebeed disabled.

```
[    5.059683] bbswitch: version 0.8
[    5.059687] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.VID_
[    5.059691] bbswitch: Found discrete VGA device 0000:02:00.0: \_SB_.PCI0.PEG_.VID_
[    5.059698] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160422/nsarguments-95)
[    5.059961] bbswitch: detected an Optimus _DSM function
[    5.059964] bbswitch: Succesfully loaded. Discrete card 0000:02:00.0 is on
```
> echo OFF > /proc/acpi/bbswitch
```
[   61.705104] vga_switcheroo: enabled
[   61.705112] bbswitch: disabling discrete graphics
[   61.705120] ACPI Warning: \_SB.PCI0.PEG.VID._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20160422/nsarguments-95)
[   61.716917] bbswitch 0000:02:00.0: Refused to change power state, currently in D0
```

So, no luck for me it seems. If you want any additional testing, want to try out a patch on this hardware etc. feel free to ping! And thanks for the try, anyway ^_^

**Comment by Lekensteyn** [comment]
`pci_set_power_state(pdev, PCI_D3hot)` should probably become `pci_set_power_state(pdev, PCI_D3cold)`. Should not make much of a difference except that maybe the "Refused to change power state" warning goes away.

Normally this should show auto/suspended for both the Nvidia and parent device:

    grep . /sys/bus/pci/devices/0000:02:00.0/{../,}power/{control,runtime_status}

For this to work, you must not have the `pcie_port_pm=off` option in your kernel command.

**Comment by abbradar** [comment]
I've come to another interesting problem -- when I'm not plugged to power adapter, X.org hangs when I have bbswitch loaded. I have discovered it only now -- bbswitch didn't work before at all because I haven't had `nvidia-drm` module blacklisted, so it got loaded and bbswitch couldn't try to disable the card. I haven't yet tried `pm-rework` branch w.r.t. this problem -- I'll do this later.

I've implemented alternative switching method proposed by @dionorgua in https://github.com/Bumblebee-Project/Bumblebee/pull/820

**Comment by y-usuzumi** [comment]
Any update?

I ran into this problem recently. My entire disk is fucked up. I was brought to the emergency mode. fsck reported millions of errors and then rendered my laptop unbootable.

**Comment by dionorgua** [comment]
If you don't use NVIDIA blob, it's already fixed. Not sure about nvidia. You just don't need to do anything except using recent kernel. nouveau is able to suspend nvidia graphics (and resume when needed)

**Comment by xobust** [comment]
I made a [fork](https://github.com/xobust/nvidia-xrun-t440p) of nvidia-xrun that automatically loads nouveau on exit to power down the GPU. This should be a ok workaround.
